Creating a getting started guide for bloom (#310)
valkey-io/valkey-doc: topics/bloomfilters.md

* If no, the user is created and the username is added to the Bloom filter.
* If yes, the app can decide to either check the main database or reject the username.

## Tutorial

**There are two ways to run valkey-bloom:**
   - **Docker**: Follow the steps here for using the [valkey bundle](valkey-bundle.md)
   - **Build from source**: Follow the [build instructions](https://github.com/valkey-io/valkey-bloom/blob/unstable/README.md#build-instructions)

### Usage example with CLI

> * Create a bloom filter of taken usernames with a 1 in 1000 chance of false positives and 10000 initial capacity.

Reviewer: ```suggestion
* Create a bloom filter to track the usernames which are taken with a 1 in 1000 chance of a false positive and 10000 as the initial capacity.
```

OK
```
* Add usernames which now have signed up to the bloom filters
```bash
# Add one user to the bloom filter
127.0.0.1:6379> BF.ADD usernames johnsmith
(integer) 1

# Add multiple users at once
127.0.0.1:6379> BF.MADD usernames JaneDoe valkeyFan bloomEnjoyer
1) (integer) 1
2) (integer) 1
3) (integer) 1
```

* Check if usernames have been taken already
```bash
# Check if as single user has been added
127.0.0.1:6379> BF.EXISTS usernames johnsmith
(integer) 1
127.0.0.1:6379> BF.EXISTS usernames fake_user
(integer) 0

# Check if multiple users are present in the bloomfilter at once
127.0.0.1:6379> BF.MEXISTS usernames johnsmith fake_user JaneDoe
1) (integer) 1
2) (integer) 0
3) (integer) 1
```

* Check how many users have been added to the filter
```bash
127.0.0.1:6379> BF.CARD usernames
(integer) 4
```
* View your bloom filter object's usage:
```bash
127.0.0.1:6379> BF.INFO usernames
 1) Capacity
 2) (integer) 10000
 3) Size
 4) (integer) 23912
 5) Number of filters
 6) (integer) 1
 7) Number of items inserted
 8) (integer) 4
 9) Error rate
10) "0.001"
11) Expansion rate
12) (integer) 2
13) Tightening ratio
14) "0.5"
15) Max scaled capacity
16) (integer) 26214300
```

> * If you know the exact number of users who will sign up, or if want to limit the number of items regardless, you can create a non scaling bloom filter. This will create a filter that has capacity for only 1000 items with a 1 in 10,000 chance of a false positive.

Reviewer: Do you mean 1 in 100,000?

The example below uses 0.00001. We can correct either one

```bash
127.0.0.1:6379> BF.RESERVE limited_users 0.00001 1000 NONSCALING
OK
```

> * If you anticipate once initial users have signed up after a while you will get a large increase in users you can set a custom expansion rate.

Reviewer: ```suggestion
* If you anticipate that, once initial users have signed up, you will get a large increase in users  after a while, you can set a custom expansion rate.
```

* If no, the user is created and the username is added to the Bloom filter.
* If yes, the app can decide to either check the main database or reject the username.

## Tutorial

### Setup

> 1. **Install Valkey**: Follow the [installation guides](installation.md) to set up Valkey on your system.

Reviewer: Why is this needed? it is not needed if someone is using Docker, right?

If you agree, lets remove this

* If no, the user is created and the username is added to the Bloom filter.
* If yes, the app can decide to either check the main database or reject the username.

> ## Tutorial

Author: I tried to follow how sets had it because they have it on top of performance and limits. But also wanted to have it after examples because that way the tutorial example makes more sense and provides context on that

To use bloom filters with Valkey, you need to:

1. **Install Valkey**: Follow the [installation guides](installation.md) to install Valkey on your system.

2. **Get the valkey-bloom module**: You have three options:
   - Use the Docker image: [valkey-extensions](https://hub.docker.com/r/valkey/valkey-bundle)
   - Download a pre-built binary from the [releases page](https://github.com/valkey-io/valkey-bloom/releases)
   - Build from source by following the [build instructions](https://github.com/valkey-io/valkey-bloom/blob/unstable/README.md#build-instructions)

Once valkey-bloom is built, you can run the Valkey server with the module loaded in two different ways, on server start up:
```bash
./valkey-server --loadmodule ./target/release/libvalkey_bloom.so
```
You can also load the valkey bloom module on an already running server by running the following command in [valkey-cli](cli.md):
```bash
127.0.0.1:6379> MODULE LOAD /path/to/libvalkey_bloom.so
```

> ### Usage example

Reviewer: ```suggestion
### CLI Usage example
```

Also, can we have a small section at the end which includes valkey-py code for bloom?

Author: I think there is meant to be a specific valkey-py page at some point where we can post the code for that. I have the code ready for that. I'd be happy to add here but there was already some concerns on the length of this

16) (integer) 26214300
```

### When to adjust default configurations

Adjusting the default configurations below can be used to update the behavior of the default bloom object created and are beneficial in different scenarios:

> - **Higher capacity (bf.bloom-capacity)**: Increase this value when you expect to store a larger number of items in all your new bloom filter objects. It increases the overall capacity for non scaling filters. For scaling filters, it increases the starting/initial capacity which reduces the need for additional scaling operations which can improve performance.

Reviewer: ```suggestion
- **Higher capacity (bf.bloom-capacity)**: Increase this value when you expect to store a larger number of items in all your new bloom filter objects. It increases the overall capacity for non scaling filters. For scaling filters, it increases the starting/initial capacity which potentially reduces the need for additional scaling operations which can improve performance.
```

- **Lower false positive rate (bf.bloom-fp-rate)**: Decrease this value when correctness is critical for your application and you want to reduce the number of false positives in all your new bloom filter objects. For example, in fraud detection or security applications where false positives are costly.

- **Higher expansion rate (bf.bloom-expansion)**: Increase this value when you want faster growth of bloom filters that will scale across all your new scaling bloom filter objects. This reduces the number of scaling operations but uses more memory quicker.

- **Lower tightening ratio (bf.bloom-tightening-ratio)**: Lower this when you want to maintain a more consistent false positive rate across multiple scaling operations for all your new bloom filter objects. (Not advisable to change this default, as it is already set to a strict value of 0.5)

> - **Fixed seed (bf.bloom-use-random-seed)**: Set to false only when you want the new bloom filter object creations to use a fixed seed for deterministic occurrence of false positives during item add/exists operations. It can be used  for testing or reproducibility.

Reviewer: ```suggestion
- **Fixed seed (bf.bloom-use-random-seed)**: Set to false only when you want the new bloom filter object creations to use a fixed seed for deterministic occurrence of false positives during item add/exists operations. It can be used for testing or reproducibility.
```

You can modify these default values using the CONFIG SET command. However, note that the effect of the configuration change is only applicable to the bloom objects created after the configuration change where the property is not specified through the command already.  

Example usage of changing all the different properties:
```bash
127.0.0.1:6379> CONFIG SET bf.bloom-fp-rate 0.001
127.0.0.1:6379> CONFIG SET bf.bloom-capacity 1000
127.0.0.1:6379> CONFIG SET bf.bloom-expansion 4
127.0.0.1:6379> CONFIG SET bf.bloom-tightening-ratio 0.6
127.0.0.1:6379> CONFIG SET bf.bloom-use-random-seed false
```

### Memory usage limit

The `bf.bloom-memory-usage-limit` configuration (default 128MB) controls the maximum memory that a single bloom filter can use:

Example usage of increasing the limit:
```bash
127.0.0.1:6379> CONFIG SET bf.bloom-memory-usage-limit 268435456
```

Having a limit on the max memory per bloom object prevents them from growing unbounded, thus maintaining server performance during serialization.

> If your use case requires bloom filters beyond this limit, you can increase this configuration value.
> 
> However, be aware it can impact overall system performance and memory availability for other operations.

Reviewer: ```suggestion
If your use case requires bloom filters with capacity beyond what this limit supports, you can increase this configuration value. However, be aware it can impact overall system performance and memory availability for other operations.
```

> When a bloom filter reaches this memory limit, any operation that would cause it to exceed the limit will fail with an error message indicating that the memory limit would be exceeded.

Reviewer: Do we want an example below?

Author: Added

    The memory usage limit per bloom filter by default is defined by the `BF.BLOOM-MEMORY-USAGE-LIMIT` module configuration which has a default value of 128 MB. If a command results in a creation / scale out causing the overall memory usage to exceed this limit, the command is rejected. This config is modifiable and can be increased as needed.

The `BF.INFO` command's `SIZE` field can be used to find out the current size of a bloom filter.

```bash

> 172.31.45.25:6379> BF.INFO validate_scale_valid SIZE

Reviewer: Should we swap this with `127.0.0.1`? :D
